### PR TITLE
feat: Support result chunks from Databricks

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

Support reading results from Databricks SQL Statement Execution API where the results are so big that they get chopped up in chunks. Read more [here](https://learn.microsoft.com/en-us/azure/databricks/sql/api/sql-execution-tutorial).

Successful run of disabled (temporarily enabled and manually deployed) domain test [here](https://github.com/Energinet-DataHub/dh3-environments/actions/runs/5256058256/jobs/9496812820).

## References

* #447

### CODE REVIEW

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SettlementReports/SettlementReportResultQueries.cs [hunk 1]
```diff
@@ -38,8 +38,14 @@ public SettlementReportResultQueries(ISqlStatementClient sqlStatementClient)
     {
         var sql = SettlementReportSqlStatementFactory.Create(gridAreaCodes, processType, periodStart, periodEnd, energySupplier);
 
-        var resultTable = await _sqlStatementClient.ExecuteSqlStatementAsync(sql).ConfigureAwait(false);
+        var rows = new List<string[]>();
+        List<string>? cols = null;
+        await foreach (var chunk in _sqlStatementClient.ExecuteAsync(sql))
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SettlementReports/SettlementReportResultQueries.cs:43]: Just a thought: can data arrive faster than it is 'consumed'? If yes, could we then run into memory issues if a lot of data arrives? This is probably not an issue for Settlement Reports (until we have basis data), so I'm mainly wondering about calculation results.

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs [hunk 2]
```diff
@@ -36,38 +36,80 @@ public SqlStatementClient(HttpClient httpClient, IOptions<DatabricksOptions> opt
...
+                await Task.Delay(TimeSpan.FromSeconds(2)).ConfigureAwait(false);
+                continue;
+            }
 
-            if (databricksSqlResponse.State is not (DatabricksSqlResponseState.Pending or DatabricksSqlResponseState.Cancelled))
+            // Handle the case where the statement did not succeed
+            if (databricksSqlResponse.State is not DatabricksSqlResponseState.Succeeded)
                 throw new DatabricksSqlException($"Unable to get calculation result from Databricks. State: {databricksSqlResponse.State}");
+
+            if (databricksSqlResponse.State == DatabricksSqlResponseState.Succeeded)
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs:85]: why is this if-statement needed when you have already handled other states in line 82

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs [hunk 3]
```diff
@@ -36,38 +36,80 @@ public SqlStatementClient(HttpClient httpClient, IOptions<DatabricksOptions> opt
...
+                yield return databricksSqlResponse.Table!;
+
+            hasMoreRows = databricksSqlResponse.HasMoreRows;
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs:88]: to me it looks like you will always return before reaching this line?

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.UnitTests/Infrastructure/SettlementReport/SettlementReportResultQueriesTests.cs [hunk 4]
```diff
@@ -66,8 +67,9 @@ public async Task GetRowsAsync_ReturnsExpectedData(Mock<ISqlStatementClient> moc
             MeteringPointType.Consumption,
             SettlementMethod.NonProfiled,
             1.234m);
-        var table = new Table(_columnNames,  new List<string[]> { row });
-        mockSqlStatementClient.Setup(s => s.ExecuteSqlStatementAsync(It.IsAny<string>())).ReturnsAsync(table);
+        var table = new TableChunk(_columnNames,  new List<string[]> { row });
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.UnitTests/Infrastructure/SettlementReport/SettlementReportResultQueriesTests.cs:70]: Aren't you missing a test that ensures that it works for two chunks?

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs:88]: Yes, `yield return ...` - that's quite different from a plain `return` statement.
See https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/yield

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs [hunk 5]
```diff
@@ -25,23 +25,31 @@ public DatabricksSqlResponse Parse(string jsonResponse)
...
         {
             case "PENDING":
-                return DatabricksSqlResponse.CreateAsPending();
+                return DatabricksSqlResponse.CreateAsPending(statementId);
             case "CANCELED":
-                return DatabricksSqlResponse.CreateAsCancelled();
+                return DatabricksSqlResponse.CreateAsCancelled(statementId);
             case "SUCCEEDED":
                 var columnNames = GetColumnNames(jsonObject);
                 var hasData = GetRowCount(jsonObject) > 0;
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:38]: I think this method returns total row count. It should probably be rows in chunk instead?

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs (hunk 2 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs]: You're right. Thx. Removed.

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs [hunk 6]
```diff
@@ -66,4 +74,9 @@ private static IEnumerable<string[]> GetDataArray(JObject responseJsonObject)
                         throw new DatabricksSqlException("Unable to retrieve 'data_array' from the responseJsonObject");
         return dataArray;
     }
+
+    private static string? GetNextChunkInternalLink(JObject responseJsonObject)
+    {
+        return responseJsonObject["result"]?["next_chunk_internal_link"]?.ToObject<string>();
```
Comment (JesperGomes) [inline source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:80]: Do we have any tests where next_chunk_internal_link exists in the result file? BTW the domain tests will not have more than one chunk.

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.UnitTests/Infrastructure/SettlementReport/SettlementReportResultQueriesTests.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.UnitTests/Infrastructure/SettlementReport/SettlementReportResultQueriesTests.cs:70]: Well there is. But there is not a test that exercises the loop being repeated - which requires two chunks _and_ an initial request, which can be either a chunk, pending or running.
Test with 3 chunks added.

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs (hunk 3 above)
Comment (JesperGomes) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/SqlStatementClient.cs:87]: Sorry - it does due to the yield. Nevermind.

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs (hunk 5 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:38]: Thx. I think it should be updated to return `.result.row_count`. Do you agree?

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs (hunk 6 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:80]: Thx. Added.
Do you have any suggestions regarding the domain test?

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs (hunk 5 above)
Comment (JesperGomes) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:38]: yes!

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs (hunk 6 above)
Comment (JesperGomes) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:80]: We could make an integration test in a scenario that triggers multiple chunks?

**Review by JesperGomes — APPROVED** [review]

File: source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs (hunk 6 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/CalculationResults/CalculationResults.Infrastructure/SqlStatements/DatabricksSqlResponseParser.cs:80]: Can you elaborate on that? How will you trigger multiple chunks? Or will you mock the SQL Statement API?

**Pull request merged**